GHC build/testsuite failures with process 1.6.13.0

There are a couple of warnings with the new refactor in #208  (promoted to errors with `-Werror`):

```
libraries/process/cbits/posix/posix_spawn.c:68:1: error:
     error: control reaches end of non-void function [-Werror=return-type]
     }
     ^
   |
68 | }
   | ^
cc1: all warnings being treated as errors
`cc' failed in phase `C Compiler'. (Exit code: 1)
libraries/process/ghc.mk:4: recipe for target 'libraries/process/dist-install/build/cbits/posix/posix_spawn.o' failed
make[1]: *** [libraries/process/dist-install/build/cbits/posix/posix_spawn.o] Error 1
make[1]: *** Waiting for unfinished jobs....
libraries/process/cbits/posix/find_executable.c: In function ‘find_executable’:
libraries/process/cbits/posix/find_executable.c:31:48: error:
     error: ‘__s’ may be used uninitialized in this function [-Werror=maybe-uninitialized]
             const int tmp_len = filename_len + 1 + strlen(path) + 1;
                                                    ^~~~~~~~~~~~
   |
31 |         const int tmp_len = filename_len + 1 + strlen(path) + 1;
   |                                                ^
libraries/process/cbits/posix/find_executable.c:26:14: error:
     note: ‘__s’ was declared here
     static char *find_in_search_path(char *search_path, const char *filename) {
                  ^~~~~~~~~~~~~~~~~~~
   |
26 | static char *find_in_search_path(char *search_path, const char *filename) {
   |              ^
cc1: all warnings being treated as errors
`cc' failed in phase `C Compiler'. (Exit code: 1)
libraries/process/cbits/posix/fork_exec.c: In function ‘do_spawn_fork’:
libraries/process/cbits/posix/fork_exec.c:222:19: error:
     error: suggest parentheses around comparison in operand of ‘&’ [-Werror=parentheses]
             if (flags & RESET_INT_QUIT_HANDLERS != 0) {
                       ^
    |
222 |         if (flags & RESET_INT_QUIT_HANDLERS != 0) {
    |                   ^
libraries/process/cbits/posix/fork_exec.c: In function ‘setup_std_handle_fork’:
libraries/process/cbits/posix/fork_exec.c:89:1: error:
     error: control reaches end of non-void function [-Werror=return-type]
     }
     ^
   |
89 | }
   | ^
cc1: all warnings being treated as errors
```

In addition, there are two testsuite failures once `-Werror` is no longer in effect.

`T3994` - `waitForProcess` now fails with `UserInterrupt` when it is expected to complete without failures

```
Wrong exit code for T3994(threaded1)(expected 0 , actual 130 )
Stdout ( T3994 ):
start
*** unexpected failure for T3994(threaded1)
```

```haskell
main :: IO ()
main = do (_,Just hout,_,p) <- createProcess (proc "./T3994app" ["start", "10000"])
                            { std_out = CreatePipe, create_group = True }
          start <- hGetLine hout
          putStrLn start
          interruptProcessGroupOf p
          t <- myThreadId
               -- timeout
          forkIO $ do
            threadDelay 5000000
            putStrLn "Interrupting a Running Process Failed"
            hFlush stdout
            killThread t
          waitForProcess p
       -- ^^^^^^^^^^^^^^^ this fails with `UserInterrupt`
          putStrLn "end"
          return ()
```

`process010` fails with a benign error message change which can probably be accepted.

```
+++ "/builds/ghc/ghc/tmp/ghctest-3wgma_aq/test   spaces/libraries/process/tests/process010.run/process010.run.stdout.normalised"	2021-07-12 19:36:34.037437016 +0000
@@ -1,4 +1,4 @@
 ExitSuccess
 ExitFailure 1
-Exc: /non/existent: rawSystem: posix_exec: illegal operation (Inappropriate ioctl for device)
+Exc: /non/existent: rawSystem: posix_exec: does not exist (No such file or directory)
 Done
```

In general, we need a better story for the GHC tests included in `process`. Ideally we would be able to run them in `process` CI as well.

Unfortunately, this means there will need to be another release soon.

/cc @bgamari

> libraries/process/cbits/posix/find_executable.c: In function ‘find_executable’:
> libraries/process/cbits/posix/find_executable.c:31:48: error:
>     error: ‘__s’ may be used uninitialized in this function [-Werror=maybe-uninitialized]

This is very strange. It looks like this `__s` symbol was introduced by a macro substitution (or similar) from the definition of `strlen`. Perhaps this is a compiler bug?

> libraries/process/cbits/posix/posix_spawn.c:68:1: error:
>      error: control reaches end of non-void function [-Werror=return-type]

This is also very strange. The `switch` indeed lacks a default case but this case should be unreachable as `hdl->behavior` is an `enum std_handle_behavior`; the switch is total with respect to the variants of this type.

Which `gcc` version are you compiling with?

> `process010` fails with a benign error message change which can probably be accepted.

Indeed it looks like we will need to normalise this difference away since it is environment-dependent.

> T3994 - `waitForProcess` now fails with `UserInterrupt` when it is expected to complete without failures

I have found a configuration where I can reproduce this failure. Thanks @wz1000!

> libraries/process/cbits/posix/find_executable.c: In function ‘find_executable’:
> libraries/process/cbits/posix/find_executable.c:31:48: error:
> error: ‘__s’ may be used uninitialized in this function [-Werror=maybe-uninitialized]

I've spent quite a while trying to understand this warning. At this point, I am quite confident that this is a `gcc` bug which sadly is only exhibited in the compiler used in Debian 9:

 * gcc 4.8.5 doesn't throw the warning
 * gcc 4.9.4 doesn't
 * gcc 6.3.0 throws the warning
 * gcc 6.5.0 doesn't
 * gcc 7.5.0 doesn't
 * gcc 8.4.0 doesn't
 * gcc 9.3.0 doesn't
 * gcc 10.3.0 doesn't
 * gcc 11.1.0 doesn't
 * clang 7.1.0 doesn't

Sadly, I think the only way to avoid this is to disable the uninitialized variable warning in the relevant part program.

I don't know if its helpfull, but hackage also fails a build on process 1.6.13.0

````
Building     process-1.6.13.0
Failed to install process-1.6.13.0
Build log ( /home/builder/.cabal/logs/ghc-8.10.2/process-1.6.13.0-8SspCePglTn1I9ZFAJtmzo.log ):
cabal: Entering directory '/tmp/cabal-tmp-20545/process-1.6.13.0'
Configuring process-1.6.13.0...
configure: WARNING: unrecognized options: --with-compiler
checking for gcc... /usr/bin/cc
checking whether the C compiler works... yes
checking for C compiler default output file name... a.out
checking for suffix of executables... 
checking whether we are cross compiling... no
checking for suffix of object files... o
checking whether we are using the GNU C compiler... yes
checking whether /usr/bin/cc accepts -g... yes
checking for /usr/bin/cc option to accept ISO C89... none needed
checking how to run the C preprocessor... /usr/bin/cc -E
checking for grep that handles long lines and -e... /bin/grep
checking for egrep... /bin/grep -E
checking for ANSI C header files... yes
checking for sys/types.h... yes
checking for sys/stat.h... yes
checking for stdlib.h... yes
checking for string.h... yes
checking for memory.h... yes
checking for strings.h... yes
checking for inttypes.h... yes
checking for stdint.h... yes
checking for unistd.h... yes
checking for pid_t... yes
checking vfork.h usability... no
checking vfork.h presence... no
checking for vfork.h... no
checking for fork... yes
checking for vfork... yes
checking for working fork... yes
checking for working vfork... (cached) yes
checking signal.h usability... yes
checking signal.h presence... yes
checking for signal.h... yes
checking sys/wait.h usability... yes
checking sys/wait.h presence... yes
checking for sys/wait.h... yes
checking fcntl.h usability... yes
checking fcntl.h presence... yes
checking for fcntl.h... yes
checking for setitimer... yes
checking for sysconf... yes
checking for execvpe... yes
checking spawn.h usability... yes
checking spawn.h presence... yes
checking for spawn.h... yes
checking for posix_spawnp... yes
checking for posix_spawn_file_actions_addchdir... no
checking whether POSIX_SPAWN_SETSID is declared... no
checking whether POSIX_SPAWN_SETSID_NP is declared... no
checking whether POSIX_SPAWN_SETPGROUP is declared... no
checking value of SIG_DFL... -1
checking value of SIG_IGN... -1
configure: creating ./config.status
config.status: creating include/HsProcessConfig.h
configure: WARNING: unrecognized options: --with-compiler
Preprocessing library for process-1.6.13.0..
Building library for process-1.6.13.0..
[1 of 5] Compiling System.Process.Common ( System/Process/Common.hs, dist/build/System/Process/Common.o, dist/build/System/Process/Common.dyn_o )
[2 of 5] Compiling System.Process.Posix ( System/Process/Posix.hs, dist/build/System/Process/Posix.o, dist/build/System/Process/Posix.dyn_o )
[3 of 5] Compiling System.Process.Internals ( System/Process/Internals.hs, dist/build/System/Process/Internals.o, dist/build/System/Process/Internals.dyn_o )
[4 of 5] Compiling System.Process   ( System/Process.hs, dist/build/System/Process.o, dist/build/System/Process.dyn_o )
[5 of 5] Compiling System.Cmd       ( System/Cmd.hs, dist/build/System/Cmd.o, dist/build/System/Cmd.dyn_o )

cbits/posix/find_executable.c:12:10: error:
     fatal error: common.h: No such file or directory
     #include "common.h"
              ^~~~~~~~~~
   |
12 | #include "common.h"
   |          ^
compilation terminated.
`cc' failed in phase `C Compiler'. (Exit code: 1)
cabal: Leaving directory '/tmp/cabal-tmp-20545/process-1.6.13.0'
````

The compiler is 8.10.2. I can build locally in Ubuntu with 8.10.4

An application I use with `process` as a dependency fails with:

```
Building library for process-1.6.13.0..
[1 of 5] Compiling System.Process.Common ( System/Process/Common.hs, dist/build/System/Process/Common.o )
[2 of 5] Compiling System.Process.Posix ( System/Process/Posix.hs, dist/build/System/Process/Posix.o )
[3 of 5] Compiling System.Process.Internals ( System/Process/Internals.hs, dist/build/System/Process/Internals.o )
[4 of 5] Compiling System.Process   ( System/Process.hs, dist/build/System/Process.o )
[5 of 5] Compiling System.Cmd       ( System/Cmd.hs, dist/build/System/Cmd.o )

cbits/posix/find_executable.c:12:10: error:
     fatal error: common.h: No such file or directory
     #include "common.h"
              ^~~~~~~~~~
   |
12 | #include "common.h"
   |          ^
```

Ubuntu 18.04, GHC 8.6.5. I think it would be beneficial to mark `process-1.6.13.0` be marked as deprecated on hackage so that cabal stops trying to use it in build plans.

--- a/cbits/posix/find_executable.c
+++ b/cbits/posix/find_executable.c
@@ -11,7 +11,7 @@
 
 #include "common.h"
 
-// the below is only necessary when we don't have execvpe.
+// the below is only necessary when we need to emulate execvpe.
 #if !defined(HAVE_execvpe)
 
 /* Return true if the given file exists and is an executable. */
@@ -28,7 +28,16 @@ static char *find_in_search_path(char *search_path, const char *filename) {
     char *tokbuf;
     char *path = strtok_r(search_path, ":", &tokbuf);
     while (path != NULL) {
+	// N.B. gcc 6.3.0, used by Debian 9, inexplicably warns that `path`
+	// may not be initialised with -Wall.  Silence this warning. See #210.
+#if defined(__GNUC__) && __GNUC__ == 6 && __GNUC_MINOR__ == 3
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wmaybe-uninitialized"
+#endif
         const int tmp_len = filename_len + 1 + strlen(path) + 1;
+#if defined(__GNUC__) && __GNUC__ == 6 && __GNUC_MINOR__ == 3
+#pragma GCC diagnostic pop
+#endif
         char *tmp = malloc(tmp_len);
         snprintf(tmp, tmp_len, "%s/%s", path, filename);
         if (is_executable(tmp)) {

--- a/cbits/posix/fork_exec.c
+++ b/cbits/posix/fork_exec.c
@@ -84,7 +84,7 @@ setup_std_handle_fork(int fd,
         if (close(b->use_pipe.parent_end) == -1) {
             child_failed(pipe, "close(parent_end)");
         }
-        break;
+	return 0;
     }
 }
 

--- a/cbits/posix/fork_exec.c
+++ b/cbits/posix/fork_exec.c
@@ -62,7 +62,6 @@ setup_std_handle_fork(int fd,
     case STD_HANDLE_CLOSE:
         if (close(fd) == -1) {
             child_failed(pipe, "close");
-            return -1;
         }
         return 0;
 

--- a/cbits/posix/fork_exec.c
+++ b/cbits/posix/fork_exec.c
@@ -224,7 +224,7 @@ do_spawn_fork (char *const args[],
 
         /* Reset the SIGINT/SIGQUIT signal handlers in the child, if requested
          */
-        if (flags & RESET_INT_QUIT_HANDLERS != 0) {
+        if ((flags & RESET_INT_QUIT_HANDLERS) != 0) {
             struct sigaction dfl;
             (void)sigemptyset(&dfl.sa_mask);
             dfl.sa_flags = 0;

--- a/cbits/posix/fork_exec.c
+++ b/cbits/posix/fork_exec.c
@@ -85,6 +85,11 @@ setup_std_handle_fork(int fd,
             child_failed(pipe, "close(parent_end)");
         }
 	return 0;
+
+    default:
+	// N.B. this should be unreachable but some compilers apparently can't
+	// see this.
+        child_failed(pipe, "setup_std_handle_fork(invalid behavior)");
     }
 }
 

--- a/cbits/posix/posix_spawn.c
+++ b/cbits/posix/posix_spawn.c
@@ -64,6 +64,13 @@ setup_std_handle_spawn (int fd,
             return -1;
         }
         return 0;
+
+    default:
+	// N.B. this should be unreachable
+	// but some compilers apparently can't
+	// see this.
+        *failed_doing = "posix_spawn_file_actions_addclose(invalid behavior)";
+        return -1;
     }
 }
 

--- a/cbits/posix/posix_spawn.c
+++ b/cbits/posix/posix_spawn.c
@@ -150,11 +150,13 @@ do_spawn_posix (char *const args[],
 #endif
     }
 
-#if defined(HAVE_POSIX_SPAWN_SETPGROUP)
     if ((flags & RUN_PROCESS_IN_NEW_GROUP) != 0) {
+#if defined(HAVE_POSIX_SPAWN_SETPGROUP)
         spawn_flags |= POSIX_SPAWN_SETPGROUP;
-    }
+#else
+	goto not_supported;
 #endif
+    }
 
     if (setup_std_handle_spawn(STDIN_FILENO,  stdInHdl,  &fa, failed_doing) != 0) {
         goto fail;

--- a/configure.ac
+++ b/configure.ac
@@ -18,9 +18,18 @@ AC_CHECK_FUNCS([execvpe])
 
 # posix_spawn checks
 AC_CHECK_HEADERS([spawn.h])
-AC_CHECK_FUNCS([posix_spawnp posix_spawn_file_actions_addchdir])
-AC_CHECK_DECLS([POSIX_SPAWN_SETSID, POSIX_SPAWN_SETSID_NP])
-AC_CHECK_DECLS([POSIX_SPAWN_SETPGROUP])
+AC_CHECK_FUNCS([posix_spawnp posix_spawn_file_actions_addchdir],[],[],[
+  #define _GNU_SOURCE
+  #include <spawn.h>
+])
+AC_CHECK_DECLS([POSIX_SPAWN_SETSID, POSIX_SPAWN_SETSID_NP],[],[],[
+  #define _GNU_SOURCE
+  #include <spawn.h>
+])
+AC_CHECK_DECLS([POSIX_SPAWN_SETPGROUP],[],[],[
+  #define _GNU_SOURCE
+  #include <spawn.h>
+])
 
 FP_CHECK_CONSTS([SIG_DFL SIG_IGN])
 

--- a/tests/all.T
+++ b/tests/all.T
@@ -38,7 +38,10 @@ test('T3994', [only_ways(['threaded1','threaded2']),
 test('T4889', normal, compile_and_run, [''])
 
 test('process009', when(opsys('mingw32'), skip), compile_and_run, [''])
-test('process010', normalise_exec, compile_and_run, [''])
+test('process010', [
+    normalise_fun(lambda s: s.replace('illegal operation (Inappropriate ioctl for device)', 'does not exist (No such file or directory)')),
+    normalise_exec
+], compile_and_run, [''])
 test('process011', when(opsys('mingw32'), skip), compile_and_run, [''])
 
 test('T8343', normal, compile_and_run, [''])

